src/client-side-metrics/metrics-collector.ts

```diff
@@ -0,0 +1,329 @@
+   * @returns {OnAttemptCompleteAttributes} An object containing the attributes for attempt metrics.
+   */
+  private getAttemptAttributes(
+    projectId: string,
+    attemptOnlyAttributes: AttemptOnlyAttributes
+  ) {
+    return Object.assign(
+      attemptOnlyAttributes,
+      this.getBasicAttributes(projectId)
+    );
+  }
+
+  /**
+   * Called when the operation starts. Records the start time.
+   */
+  onOperationStart() {
+    this.operationStartTime = this.dateProvider.getDate();
+  }
+
+  /**
+   * Called when an attempt (e.g., an RPC attempt) completes. Records attempt latencies.
+   * @param {AttemptOnlyAttributes} info Information about the completed attempt.
+   */
+  onAttemptComplete(info: OnAttemptCompleteInfo) {
+    this.attemptCount++;
+    const endTime = this.dateProvider.getDate();
+    const projectId = this.projectId;
+    if (projectId && this.attemptStartTime) {
+      const attributes = this.getAttemptAttributes(projectId, info);
+      const totalTime = endTime.getTime() - this.attemptStartTime.getTime();
+      this.metricsHandlers.forEach(metricsHandler => {
+        if (metricsHandler.onAttemptComplete) {
+          metricsHandler.onAttemptComplete(
+            {
+              attemptLatency: totalTime,
+              serverLatency: this.serverTime ?? undefined,
+              connectivityErrorCount: info.connectivityErrorCount,
+              firstResponseLatency: this.firstResponseLatency ?? undefined,
+            },
+            attributes
+          );
+        }
+      });
+    }
+  }
+
+  /**
+   * Called when a new attempt starts. Records the start time of the attempt.
+   */
+  onAttemptStart() {
+    this.attemptStartTime = this.dateProvider.getDate();
+    this.serverTime = null;
+    this.serverTimeRead = false;
+    this.firstResponseLatency = null;
+    this.receivedFirstResponse = false;
+  }
+
+  /**
+   * Called when the first response is received. Records first response latencies.
+   * @param {string} finalOperationStatus The final status of the operation.
+   */
+  onResponse() {
+    const endTime = this.dateProvider.getDate();
+    const projectId = this.projectId;
+    if (projectId && this.attemptStartTime) {
+      const totalTime = endTime.getTime() - this.attemptStartTime.getTime();
+      if (!this.receivedFirstResponse) {
+        this.receivedFirstResponse = true;
+        this.firstResponseLatency = totalTime;
+      }
+    }
+  }
+
+  /**
+   * Called when an operation completes (successfully or unsuccessfully).
+   * Records operation latencies, retry counts, and connectivity error counts.
+   * @param {OperationOnlyAttributes} info Information about the completed operation.
+   */
+  onOperationComplete(info: OperationOnlyAttributes) {
+    const endTime = this.dateProvider.getDate();
+    const projectId = this.projectId;
+    if (projectId && this.operationStartTime) {
+      const totalTime = endTime.getTime() - this.operationStartTime.getTime();
```

Comment:
consider some kind of protection to make sure this doesn't get called multiple times

In Python, we used a state machine to track the lifecycle

Reply:
A state machine is now in the PR